Fix usage of MustCompile

While reviewing some code I've noticed we call `regexp.MustCompile` in places where we should not panic.

```
$ git grep MustCompile -- pkg
pkg/build/strategies/layered/layered.go:        tarHandler.SetExclusionPattern(regexp.MustCompile(config.ExcludeRegExp))
pkg/build/strategies/onbuild/entrypoint.go:     regexp.MustCompile(`^run(\.sh)?$`),
pkg/build/strategies/onbuild/entrypoint.go:     regexp.MustCompile(`^start(\.sh)?$`),
pkg/build/strategies/onbuild/entrypoint.go:     regexp.MustCompile(`^exec(\.sh)?$`),
pkg/build/strategies/onbuild/entrypoint.go:     regexp.MustCompile(`^execute(\.sh)?$`),
pkg/build/strategies/sti/sti.go:        tarHandler.SetExclusionPattern(regexp.MustCompile(config.ExcludeRegExp))
pkg/docker/util.go:var dockerLineDelim = regexp.MustCompile(`[\t\v\f\r ]+`)
pkg/scm/git/git.go:var gitSSHURLUser = regexp.MustCompile(`^([\w\-_\.]+)$`)
pkg/scm/git/git.go:var gitSSHURLIPv4 = regexp.MustCompile(`^([\w\.\-_]+)$`)         // should cover textual hostnames and x.x.x.x
pkg/scm/git/git.go:var gitSSHURLIPv6 = regexp.MustCompile(`^(\[[a-fA-F0-9\:]+\])$`) // should cover [hex:hex ... :hex]
pkg/scm/git/git.go:var gitSSHURLPathRef = regexp.MustCompile(`^([\w\.\-_+\/\\]+)$`)
pkg/tar/tar.go:var DefaultExclusionPattern = regexp.MustCompile("((^\\.git\\/)|(\\/.git\\/)|(\\/.git$))")
pkg/tar/tar_test.go:    th.SetExclusionPattern(regexp.MustCompile("test3.txt"))
pkg/tar/tar_test.go:    th.SetExclusionPattern(regexp.MustCompile(""))
```

As explained in https://talks.golang.org/2014/readability.slide#9, `regexp.MustCompile` should be used only in initialization (package var or init()).

Usages like in `pkg/build/strategies/layered/layered.go` and `pkg/build/strategies/sti/sti.go` may cause unhandled panics.

I realized that this can actually lead to easy to reproduce panics, by simply passing a malformed regexp to the `--exclude` flag of `s2i build`:

```
[vagrant@localhost source-to-image]$ _output/local/bin/linux/amd64/s2i build --exclude '[' . centos/ruby-22-centos7 hello-world
panic: regexp: Compile(`[`): error parsing regexp: missing closing ]: `[`

goroutine 1 [running]:
panic(0x7a57e0, 0xc42030e7f0)
        /usr/lib/golang/src/runtime/panic.go:500 +0x1a1
regexp.MustCompile(0x7ffee752b4f8, 0x1, 0x0)
        /usr/lib/golang/src/regexp/regexp.go:237 +0x17a
github.com/openshift/source-to-image/pkg/build/strategies/sti.New(0xc4202a6000, 0xc1fe00, 0xc4200110e0, 0x0, 0x0, 0xc420246f00, 0xc420012e18, 0x6)
        /home/vagrant/src/github.com/openshift/source-to-image/_output/local/go/src/github.com/openshift/source-to-image/pkg/build/strategies/sti/sti.go:110 +0x23b
github.com/openshift/source-to-image/pkg/build/strategies.Strategy(0xc4202a6000, 0x0, 0x0, 0x5, 0xc4200110a0, 0x15, 0x0, 0xc42001ad20, 0x23, 0x0, ...)
        /home/vagrant/src/github.com/openshift/source-to-image/_output/local/go/src/github.com/openshift/source-to-image/pkg/build/strategies/strategies.go:51 +0x171
github.com/openshift/source-to-image/pkg/build/strategies.GetStrategy(0xc4202a6000, 0x0, 0x0, 0xc42026b930, 0x1, 0x1, 0x52620172f59cdd5d, 0x0, 0x811080)
        /home/vagrant/src/github.com/openshift/source-to-image/_output/local/go/src/github.com/openshift/source-to-image/pkg/build/strategies/strategies.go:16 +0x47
main.newCmdBuild.func1(0xc42009b200, 0xc42021abe0, 0x3, 0x5)
        /home/vagrant/src/github.com/openshift/source-to-image/_output/local/go/src/github.com/openshift/source-to-image/cmd/s2i/main.go:155 +0x703
github.com/openshift/source-to-image/vendor/github.com/spf13/cobra.(*Command).execute(0xc42009b200, 0xc42021ab90, 0x5, 0x5, 0xc42009b200, 0xc42021ab90)
        /home/vagrant/src/github.com/openshift/source-to-image/_output/local/go/src/github.com/openshift/source-to-image/vendor/github.com/spf13/cobra/command.go:603 +0x439
github.com/openshift/source-to-image/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc42009ad80, 0xc420191eb0, 0x0, 0xffffffffffffffff)
        /home/vagrant/src/github.com/openshift/source-to-image/_output/local/go/src/github.com/openshift/source-to-image/vendor/github.com/spf13/cobra/command.go:689 +0x367
github.com/openshift/source-to-image/vendor/github.com/spf13/cobra.(*Command).Execute(0xc42009ad80, 0xc420191ea8, 0x1)
        /home/vagrant/src/github.com/openshift/source-to-image/_output/local/go/src/github.com/openshift/source-to-image/vendor/github.com/spf13/cobra/command.go:648 +0x2b
main.main()
        /home/vagrant/src/github.com/openshift/source-to-image/_output/local/go/src/github.com/openshift/source-to-image/cmd/s2i/main.go:617 +0x669
```

--- a/pkg/build/strategies/layered/layered.go
+++ b/pkg/build/strategies/layered/layered.go
@@ -42,12 +42,19 @@ type Layered struct {
 
 // New creates a Layered builder.
 func New(config *api.Config, fs util.FileSystem, scripts build.ScriptsHandler, overrides build.Overrides) (*Layered, error) {
+	excludePattern, err := regexp.Compile(config.ExcludeRegExp)
+	if err != nil {
+		return nil, err
+	}
+
 	d, err := docker.New(config.DockerConfig, config.PullAuthentication)
 	if err != nil {
 		return nil, err
 	}
+
 	tarHandler := tar.New(fs)
-	tarHandler.SetExclusionPattern(regexp.MustCompile(config.ExcludeRegExp))
+	tarHandler.SetExclusionPattern(excludePattern)
+
 	return &Layered{
 		docker:  d,
 		config:  config,

--- a/pkg/build/strategies/layered/layered_test.go
+++ b/pkg/build/strategies/layered/layered_test.go
@@ -6,10 +6,12 @@ import (
 	"os"
 	"path/filepath"
 	"regexp"
+	"regexp/syntax"
 	"strings"
 	"testing"
 
 	"github.com/openshift/source-to-image/pkg/api"
+	"github.com/openshift/source-to-image/pkg/build"
 	"github.com/openshift/source-to-image/pkg/docker"
 	"github.com/openshift/source-to-image/pkg/test"
 )
@@ -183,5 +185,15 @@ func TestBuildErrorOnBuildBlocked(t *testing.T) {
 	_, err := l.Build(l.config)
 	if err == nil || !strings.Contains(err.Error(), "builder image uses ONBUILD instructions but ONBUILD is not allowed") {
 		t.Errorf("expected error from onbuild due to blocked ONBUILD, got: %v", err)
+	}
+}
+
+func TestNewWithInvalidExcludeRegExp(t *testing.T) {
+	_, err := New(&api.Config{
+		DockerConfig:  docker.GetDefaultDockerConfig(),
+		ExcludeRegExp: "[",
+	}, nil, nil, build.Overrides{})
+	if syntaxErr, ok := err.(*syntax.Error); ok && syntaxErr.Code != syntax.ErrMissingBracket {
+		t.Errorf("expected regexp compilation error, got %v", err)
 	}
 }

--- a/pkg/build/strategies/sti/sti.go
+++ b/pkg/build/strategies/sti/sti.go
@@ -86,6 +86,11 @@ type STI struct {
 // be used for the case that the base Docker image does not have 'tar' or 'bash'
 // installed.
 func New(config *api.Config, fs util.FileSystem, overrides build.Overrides) (*STI, error) {
+	excludePattern, err := regexp.Compile(config.ExcludeRegExp)
+	if err != nil {
+		return nil, err
+	}
+
 	docker, err := dockerpkg.New(config.DockerConfig, config.PullAuthentication)
 	if err != nil {
 		return nil, err
@@ -107,7 +112,7 @@ func New(config *api.Config, fs util.FileSystem, overrides build.Overrides) (*ST
 		fs,
 	)
 	tarHandler := tar.New(fs)
-	tarHandler.SetExclusionPattern(regexp.MustCompile(config.ExcludeRegExp))
+	tarHandler.SetExclusionPattern(excludePattern)
 
 	builder := &STI{
 		installer:              inst,

--- a/pkg/build/strategies/sti/sti_test.go
+++ b/pkg/build/strategies/sti/sti_test.go
@@ -6,6 +6,7 @@ import (
 	"io"
 	"path/filepath"
 	"reflect"
+	"regexp/syntax"
 	"strings"
 	"testing"
 
@@ -903,5 +904,15 @@ func TestCleanup(t *testing.T) {
 		} else if !p && removedDir == "" {
 			t.Errorf("Expected working directory to be removed, but it was preserved.")
 		}
+	}
+}
+
+func TestNewWithInvalidExcludeRegExp(t *testing.T) {
+	_, err := New(&api.Config{
+		DockerConfig:  docker.GetDefaultDockerConfig(),
+		ExcludeRegExp: "(",
+	}, nil, build.Overrides{})
+	if syntaxErr, ok := err.(*syntax.Error); ok && syntaxErr.Code != syntax.ErrMissingParen {
+		t.Errorf("expected regexp compilation error, got %v", err)
 	}
 }

